src/core/mac/link_raw.cpp

```diff
@@ -224,117 +238,111 @@ void LinkRaw::InvokeReceiveDone(otRadioFrame *aFrame, otError aError)
     }
 }
 
-otError LinkRaw::Transmit(otRadioFrame *aFrame, otLinkRawTransmitDone aCallback)
+void LinkRaw::StartTransmit(void)
 {
-    otError error = OT_ERROR_INVALID_STATE;
+#if OPENTHREAD_CONFIG_ENABLE_SOFTWARE_CSMA_BACKOFF
+    if ((mRadioCaps & OT_RADIO_CAPS_CSMA_BACKOFF) == 0 && mTransmitFrame->mInfo.mTxInfo.mCsmaCaEnabled)
+    {
+        StartCsmaBackoff();
+    }
+    else
+#endif // OPENTHREAD_CONFIG_ENABLE_SOFTWARE_CSMA_BACKOFF
+    {
+        mPendingOperations |= kOperationTransmitData;
+        mOperationTask.Post();
+    }
+}
 
-    if (mEnabled)
+void LinkRaw::TransmitNow(void)
+{
+    otError error = otPlatRadioTransmit(&GetInstance(), mTransmitFrame);
+
+    if (error != OT_ERROR_NONE)
     {
-        mTransmitDoneCallback = aCallback;
+        InvokeTransmitDone(mTransmitFrame, NULL, error);
+    }
+}
+
+otError LinkRaw::Transmit(otRadioFrame *aFrame, otLinkRawTransmitDone aCallback)
+{
+    otError error = OT_ERROR_NONE;
+
+    assert(mTransmitFrame == aFrame);
+    VerifyOrExit(mEnabled, error = OT_ERROR_INVALID_STATE);
+
+    mTransmitDoneCallback = aCallback;
 
 #if OPENTHREAD_CONFIG_ENABLE_SOFTWARE_RETRANSMIT
+    if ((mRadioCaps & OT_RADIO_CAPS_TRANSMIT_RETRIES) == 0)
+    {
         mTransmitRetries = 0;
         mCsmaBackoffs    = 0;
+    }
 #endif // OPENTHREAD_CONFIG_ENABLE_SOFTWARE_RETRANSMIT
 
-#if OPENTHREAD_CONFIG_ENABLE_SOFTWARE_CSMA_BACKOFF
-        if (aFrame->mInfo.mTxInfo.mCsmaCaEnabled)
-        {
-            // Start the transmission backoff logic
-            StartCsmaBackoff();
-            error = OT_ERROR_NONE;
-        }
-        else
-        {
-#endif // OPENTHREAD_CONFIG_ENABLE_SOFTWARE_CSMA_BACKOFF
-            error = otPlatRadioTransmit(&mInstance, aFrame);
-#if OPENTHREAD_CONFIG_ENABLE_SOFTWARE_CSMA_BACKOFF
-        }
-#endif // OPENTHREAD_CONFIG_ENABLE_SOFTWARE_CSMA_BACKOFF
-    }
+    StartTransmit();
 
+exit:
     return error;
 }
 
 void LinkRaw::InvokeTransmitDone(otRadioFrame *aFrame, otRadioFrame *aAckFrame, otError aError)
 {
-    otLogDebgPlat("LinkRaw Transmit Done (err=0x%x)", aError);
+    if (aError == OT_ERROR_NONE)
+    {
+        otLogDebgPlat("LinkRaw Transmit Done: %s", otThreadErrorToString(aError));
+    }
+    else
+    {
+        otLogWarnPlat("LinkRaw Transmit Done: %s", otThreadErrorToString(aError));
+    }
+
+    assert(aFrame == mTransmitFrame);
 
 #if OPENTHREAD_CONFIG_ENABLE_SOFTWARE_ACK_TIMEOUT
-    mTimer.Stop();
+    if ((mRadioCaps & OT_RADIO_CAPS_ACK_TIMEOUT) == 0)
+    {
+        mTimer.Stop();
+    }
 #endif // OPENTHREAD_CONFIG_ENABLE_SOFTWARE_ACK_TIMEOUT
 
 #if OPENTHREAD_CONFIG_ENABLE_SOFTWARE_RETRANSMIT
```

Comment:
What happens if radio provides support for CSMA backoff but not retries and we have `SOFTWARE_RETRANSMIT` enabled?
In that case, if the error `CHANNEL_ACCESS_FAILURE` we should not do `mCsmaBackOff++` and the CSMA retry and directly go to check for retx. 

I suggest keeping the logic separate, i.e., we first check if sw CSMA backoff is needed (`SOFTWARE_CSMA_BACKOFF` is enabled and radio does no provide it) and handle CSMA retry and then check if retx is needed and handle it (sort of like the `mac.cpp` implementation [here](https://github.com/openthread/openthread/blob/master/src/core/mac/mac.cpp#L1346))